### PR TITLE
Requestify Closure Capture Computation

### PR DESCRIPTION
**Pull request by CodaFi**
Capture info is now computed on-demand (assuming a function body is available).  As a consequence, there is no longer a need for the type checker to track closures, and less of a reason for it to track function decls.  The queue of closures is now gone and the type checker is a bit simpler internally.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
apple/swift-lldb#1941
@swift-ci Please smoke test

File: lib/AST/TypeCheckRequests.cpp [hunk 1]
```diff
@@ -846,6 +846,8 @@ Optional<CaptureInfo> ComputeCaptureInfoRequest::getCachedResult() const {
 }
 
 void ComputeCaptureInfoRequest::cacheResult(CaptureInfo value) const {
+  // FIXME: This is quite messy.  The layout of CaptureInfo could use some
```
Comment (slavapestov) [inline lib/AST/TypeCheckRequests.cpp]: It's also too large, with a number of fields that in most functions, are null. The capture info should be a single ArrayRef, with the dynamic self and generic captures cases encoded as elements of the array.

**Comment by CodaFi** [comment]
Going to put this on the back burner. There’s structural issues with this being in Sema at all.

**Pull request closed without merging**